x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/aggs/kstest/BucketCountKSTestAggregator.java

```diff
@@ -0,0 +1,253 @@
+        // gives us unreliable statistics.
+        // TODO bootstrap to support sparse data
+        if (nSamples < MINIMUM_NUMBER_OF_DOCS) {
+            return alternatives.stream().map(Alternative::toString).collect(Collectors.toMap(Function.identity(), a -> Double.NaN));
+        }
+        DoubleArray.divMut(fY, fY[fY.length - 1]);
+        final double[] monotonic = LongStream.range(1, fY.length + 1).mapToDouble(Double::valueOf).toArray();
+
+        // fast path, since our sample is the entire cdf space, don't worry about multiple passes on sampling
+        if (nSamples >= samplingMethod.cdfPoints().length) {
+            return ksTest(nSamples, fX, fY, monotonic, samplingMethod.cdfPoints(), alternatives);
+        }
+
+        Map<String, Double> result = Stream.generate(() -> ksTest(nSamples, fX, fY, monotonic, samplingMethod.cdfPoints(), alternatives))
+            .limit(NUM_ITERATIONS)
+            .reduce(new HashMap<>(), (memo, v) -> {
+                v.forEach(
+                    (alternative, ksTestValue) -> memo.merge(alternative, ksTestValue, (v1, v2) -> v1 + (v2 == 0.0 ? 0.0 : Math.log(v2)))
+                );
+                return memo;
+            });
+        alternatives.stream()
+            .map(Alternative::toString)
+            .forEach(a -> result.put(a, Math.min(1.0, Math.max(Math.exp(result.get(a) / NUM_ITERATIONS), 0.0))));
+        return result;
+    }
+
+    private static Map<String, Double> ksTest(
+        int nSamples,
+        double[] fX,
+        double[] fY,
+        double[] monotonic,
+        double[] cdfPoints,
+        EnumSet<Alternative> alternatives
+    ) {
+        int[] samples = sampleOf(cdfPoints.length, nSamples);
+        double[] x = new double[samples.length];
+        double[] y = new double[samples.length];
+        int index = 0;
+        for (int i : samples) {
+            double f = cdfPoints[i];
+            x[index] = interpolate(fX, monotonic, f);
+            y[index] = interpolate(fY, monotonic, f);
+            ++index;
+        }
+        // These arrays should typically already be sorted
+        // But, depending on the sampling methodology, they could be monotonically DECREASING instead of INCREASING
+        // calling `sort` to capture this situation reliably
+        Arrays.sort(x);
+        Arrays.sort(y);
+
+        Map<String, Double> results = new HashMap<>();
+        // The following approximation is the same as eq 5.3 from:
+        // J. L. Hodges
+        // "The significance probability of the smirnov two-sample test," Arkiv för Matematik, Ark. Mat. 3(5), 469-486, (10 Jan. 1958)
+        //
+        // The other option here for an approximation would be Donald Knuths provided through:
+        // Dₙ,ₘ>c(α)*√((n+m)/(n*m)).
+        // And c(α)=√(−ln(α * 0.5) * 0.5)
+        // Where Dₙ,ₘ is our statistic
+        // Then solve for α
+        // But I am not 100% sure which to choose in this case.
```

Comment:
This distribution is for the usual K-S test statistic not the sided versions. For small p-values the sided versions are very nearly half the two-sided one. I'll dig around for a similar asymptotic approximation for these alternatives.

Reply:
Ah, yeah, hodges approximation is for the sided-statistic. So, we will have to have something else for two_sided.